### DISCUSSION TITLE
[Java] - G.C 동작 원리

### DISCUSSION THREAD

**Discussion by Irisation23** [cs-collections/cs-collections-discussion#42] [Java]
# 1. 가비지 컬렉션(Garbage Collection)

- 자바의 메모리 관리 방법 중 하나임.
- `JVM`의 `Heap 영역`에서 `동적으로 할당했던 메모리 영역` 중 필요 없게 된 메모리 영역을 주기적으로 삭제하는 프로세스를 말함.
- 메모리 관리, 메모리 누수(Memory Leak) 문제를 개발자가 신경 쓰지 않아도 된다.

## 1.1 가비지 컬렉션의 단점 ✨

1. 개발자가 메모리 해제 시점을 정확히 알 수 없다.
2. GC 동작시 다른 동작을 멈추기 때문에 오버헤드가 발생한다.

<details>
<summary>오버헤드 란?</summary>
<div markdown="1">
<img width="964" alt="image" src="https://user-images.githubusercontent.com/53285909/214313124-6bad2af8-7d34-4f56-95f4-f30d92468b02.png">

</div>
</details>

## 1.2 가바지 컬렉션의 수거 대상 

- 객체들은 Heap영역에서 생성되고 Method Area이나 Stack Area등 Root Area에서는 Heap Area에 생성된 객체의 주소만 참조하는 형식으로 구성된다.
- 객체 참조가 Unreachable 하다면 가바지 컬렉터가 이를 수거 한다.

## 1.3 가비지 컬렉션의 동작 과정

<img width="682" alt="image" src="https://user-images.githubusercontent.com/53285909/214876409-d01d20a9-35e6-4742-a8cc-ae4722809655.png">

> Heap Area는 효율적인 GC를 위해 Eden, Survaival, Old Generation으로 나뉜다.

<img width="657" alt="image" src="https://user-images.githubusercontent.com/53285909/214883073-bc00256c-3f6d-412f-b3c2-ed04acd8136a.png">

1. 객체가 처음 생성되고 Heap 영역의 `Eden`에 `age-bit 0` 으로 할당된다. 이 age-bit는 `Minor GC`에서 살아 남을 때마다 1씩 증가하게 된다.

<img width="383" alt="image" src="https://user-images.githubusercontent.com/53285909/214884335-5d26a961-6b91-41f6-a275-4dda712d310b.png">

2. 시간이 지나 Heap Area의 Eden 영역에 객체가 다 쌓이게 되면 Minor GC가 한번 일어나게 되고 참조 정도에 따라 Servivor0 영역으로 이동하거나 회수된다.

<img width="513" alt="image" src="https://user-images.githubusercontent.com/53285909/214885038-0c2275af-7868-482b-bcde-0f92c95979dd.png">

3. Eden 영역에 신규 객체가 생성된다. Eden 영역에 객체가 쌓이게 되면 Young Generation(Eden + Servivor) 영역에 있는 객체들을 비어있는 Survival인 Survival1 영역에 이동하고 살아남은 모든 객체들은 age가 1씩 증가한다. 

<img width="565" alt="image" src="https://user-images.githubusercontent.com/53285909/214885452-e858ff2f-7b1a-4944-8649-360e3d74e4c5.png">

4. Eden 영역에 신규 객체들로 가득 차게 되면 다시한번 minor GC가 일어나고 Young Generation(Eden +  Survivor) 영역에 있는 객체들을 비어있는 Survival인 Survival0으로 이동시킨 뒤 age 1 증가 시킨다. 해당 과정은 계속해서 반복된다. 

<img width="633" alt="image" src="https://user-images.githubusercontent.com/53285909/214886106-feb35d28-7f21-4ccb-9dec-57151ed0d509.png">

5. 해당 과정을 반복하다 보면 age bit가 특정 숫자 이상으로 되는 경우가 발생한다. 이때 JVM에서 설정해 놓은 age bit에 도달하게 되면 오랫동안 쓰일 객체라고 판단하고 Old generation 영역으로 이동시킨다. 해당 과정을 `프로모션(Promotion)`이라고 한다. 

<img width="374" alt="image" src="https://user-images.githubusercontent.com/53285909/214886818-d3b777c6-512e-4115-846e-03d526f35432.png">

6. 시간이 지나 Old 영역에 할당된 메모리가 허용치를 넘게 되면, Old 영역에 있는 모든 객체들을 검사하여 참조되지 않는 객체들을 한꺼번에 삭제하는 GC가 실행된다. 이렇게 Old generation영역의 메모리를 회수하는 GC를 `Major GC`라고 한다. Major GC는 오래 걸리는 작업이고, 이때 GC를 실행하는 스레드를 제외한 모든 스레드는 작업을 멈추게 된다. 이를 `Stop-the-World`라고 한다. 해당 작업이 잦으면 프로그램 성능에 문제가 될 수 있다.

# 2. GC 정리 📚
- GC는 동적 메모리 영역 중 필요 없게된 메모리 영역을 주기적으로 삭제하는 프로세스를 말한다.
- GC대상은 Heap 영역 안에 있는 객체중 참조를 잃은 객체 이다.(참조 되지 않는 객체)
- GC의 동작은 Eden과 Survivor 영역에서 Minor GC가 이루어진 뒤 Old generation에 프로모션 된 후 Major GC를 통해 동작하게 된다.
- Major GC는 GC이외의 스레드를 멈추기 때문에 성능 이슈가 생길 수 있다.

> 참조
> https://coding-factory.tistory.com/829

**Comment by JoisFe** [comment]
GC의 구체적인 동작원리를 덕분에 알게 되었습니다. 감사합니다^^
